rpcs3/Emu/RSX/Overlays/overlay_message.cpp

```diff
@@ -6,88 +6,178 @@ namespace rsx
+		}
+
+		void message_item::set_pos(u16 _x, u16 _y)
+		{
+			rounded_rect::set_pos(_x, _y);
+			m_text.set_pos(_x + m_margin, y + m_margin);
+
+			if (m_icon)
 			{
-				return {};
+				m_icon->set_pos(m_icon->x, m_text.y);
 			}
+		}
 
-			auto cr = m_text.get_compiled();
-			m_fade_animation.apply(cr);
+		compiled_resource& message_item::get_compiled()
+		{
+			auto& current_animation = m_fade_in_animation.active
+				? m_fade_in_animation
+				: m_fade_out_animation;
 
-			return cr;
+			if (!m_processed || !current_animation.active)
+			{
+				compiled_resources = {};
+				return compiled_resources;
+			}
+
+			// Disable caching
+			is_compiled = false;
+
+			compiled_resources = rounded_rect::get_compiled();
+			compiled_resources.add(m_text.get_compiled());
+			if (m_icon)
+			{
+				compiled_resources.add(m_icon->get_compiled());
+			}
+
+			current_animation.apply(compiled_resources);
+			return compiled_resources;
 		}
 
-		void message_item::update(usz index, u64 time)
+		void message_item::update(usz index, u64 time, u16 y_offset)
 		{
 			if (m_cur_pos != index)
 			{
 				m_cur_pos = index;
-				m_text.set_pos(10, static_cast<u16>(index * 18));
+				set_pos(10, y_offset);
+			}
+
+			if (!m_processed)
+			{
+				m_expiration_time = get_expiration_time(m_visible_duration);
 			}
 
-			if ((m_expiration_time - time) < 2'000'000)
+			if (m_fade_in_animation.active)
+			{
+				m_fade_in_animation.update(rsx::get_current_renderer()->vblank_count);
+			}
+			else if (time + 2'000'000 > m_expiration_time)
 			{
-				m_fade_animation.update(rsx::get_current_renderer()->vblank_count);
+				m_fade_out_animation.update(rsx::get_current_renderer()->vblank_count);
 			}
 
 			m_processed = true;
 		}
 
-		void message::update()
+		void message::update_queue(std::deque<message_item>& vis_set, std::deque<message_item>& ready_set, message_pin_location origin)
 		{
-			if (!visible)
+			const u64 cur_time = rsx::uclock();
+
+			while (!vis_set.empty() && vis_set.front().get_expiration() < cur_time)
 			{
-				return;
+				vis_set.pop_front();
 			}
 
-			std::lock_guard lock(m_mutex_queue);
-			u64 cur_time = get_system_time();
-
-			while (!m_queue.empty() && m_queue.front().get_expiration() < cur_time)
+			while (vis_set.size() < max_visible_items && !ready_set.empty())
 			{
-				m_queue.pop_front();
+				vis_set.emplace_back(std::move(ready_set.front()));
+				ready_set.pop_front();
 			}
 
-			if (m_queue.empty())
+			if (vis_set.empty() && ready_set.empty())
 			{
-				visible = false;
 				return;
 			}
 
-			usz index = 0;
+			// Render reversed list. Oldest entries are furthest from the border
+			u16 y_offset = 8, margin = 4, index = 0;
+			for (auto it = vis_set.rbegin(); it != vis_set.rend(); ++it, ++index)
+			{
+				if (origin == message_pin_location::top) [[ likely ]]
+				{
+					it->update(index, cur_time, y_offset);
+					y_offset += (margin + it->h);
+				}
+				else
+				{
+					y_offset += (margin + it->h);
```

Comment:
why do we add the margin beforehand? we already have the 8 pixel offset at the first message.

Reply:
Looks nicer with a 16px offset vs 8. The same config up top looks awful though.